Som-Energia/heman
# NEW: timescale migration

## Description

This PR add a new implementation of the drivers to support direct timescale access reusing [`infoenergia-api` strategy](https://github.com/Som-Energia/infoenergia-api/pull/60).

## Changes
- New plugable and configurable curve backends
- New curve backend for Timescale based curves
- Fix: curve time handling and conversion, now curves properly
  start and end in 00:00 local time and timestamps base on UTC Epoch
- Upgrade notes:
    - Need to set the environ `POSTGRESQL_URI` pointing to ERP database
    for the timescale curve backend.
    - You can set environ `DEFAULT_CURVE_BACKEND` to set the default
    to either `mongo` or `timescale`, default `mongo`
    - You can set environ `CURVE_BACKENDS` to a json dict, mapping
    concrete curves (`tg_f1`...) to a different backend.

## Observations
This branch is born from [NEW_pvcalculator](https://github.com/Som-Energia/heman/tree/NEW_pvcalculator) which fixes some application errors but when it is deprecated we delete the code involved.

## Please, review
- datetime curve alignment have been changed to respect DST
- how backend configuration have been implemented
- datetime utilities: `datetimeutils.py`

## How to check the new features

- heman-test deployed pointing to serp mongo and postgresql like ov-test.
- Enter https://ovtest.somenergia.coop with the user you want to test in order to ensure an updated token
- Go to infoenergia curves and inspect the queries to heman-test to get the token.
- Other ways of getting the token:
    - running the infoenergia-dashboard/scrips/generate_env.py script
    - searching in mongo in the token collection
    - searching in erp res_partner.infoenergia_token
- If there is no curves in mongo you have to import then from production into `serp01` mongo (just for mongo curves, not timescale ones):
    - Use this script launcher: http://scriptlauncher.somenergia.lan/runner/CopyCurveToTesting
- If you want to run the api directy, there is a convenience script with plugable datafiles:
    - Go somenergia-back2backdata/heman/testdata
   - `./smoketest_curves.sh case-f1-p1.env`
   - Create any cases you need as `case-*.env` files
   - Importantly, one of the existing cases has no f5d curves and has to use the the timescale backend for f1's.


## Deploy notes
* There are **new Python requirements**, don't forget to install them via `./setup.py develop` or `pip install -r requirements.txt`
* There are a **new environment variable** `POSTGRESQL_URI` to be set up to get the application working with postgresql
